pkg/apis/v1alpha1/pipeline.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2021 VMware
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +versionName=v1alpha1
+// +groupName=carto.run
+// +kubebuilder:object:generate=true
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	PipelineReady    = "Ready"
+	RunTemplateReady = "RunTemplateReady"
+)
+
+const (
+	ReadyRunTemplateReason                            = "Ready"
+	NotFoundRunTemplateReason                         = "RunTemplateNotFound"
+	StampedObjectRejectedByAPIServerRunTemplateReason = "StampedObjectRejectedByAPIServer"
+	TemplateStampFailureRunTemplateReason             = "TemplateStampFailure"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+type Pipeline struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              PipelineSpec   `json:"spec"`
+	Status            PipelineStatus `json:"status,omitempty"`
+}
+
+type PipelineStatus struct {
+	ObservedGeneration int64              `json:"observedGeneration,omitempty"`
+	Conditions         []metav1.Condition `json:"conditions,omitempty"`
+}
+
+type PipelineSpec struct {
+	// +kubebuilder:validation:Required
+	RunTemplateRef TemplateReference `json:"runTemplateRef"`
+}
+
+type TemplateReference struct {
+	// +kubebuilder:validation:Enum=RunTemplate
+	Kind string `json:"kind"`
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name"`
+	// +kubebuilder:validation:MinLength=1
+	Namespace string `json:"namespace"`
```

Comment:
in the same line of thinking of the comment about `kind`, I think we could make the ux better here by making the namespace not required, but rather optional, defaulting to  the same namespace as the pipeline

```suggestion
	Namespace string `json:"namespace,omitempty"`
```